src/diffusers/loaders.py

```diff
@@ -85,7 +85,49 @@ def __init__(self, regular_linear_layer, lora_scale=1, network_alpha=None, rank=
 
         self.lora_scale = lora_scale
 
+    def _fuse_lora(self):
+        if self.lora_linear_layer is None:
+            return
+
+        dtype, device = self.regular_linear_layer.weight.data.dtype, self.regular_linear_layer.weight.data.device
+        logger.info(f"Fusing LoRA weights for {self.__class__}")
+
+        w_orig = self.regular_linear_layer.weight.data.float()
+        w_up = self.lora_linear_layer.up.weight.data.float()
+        w_down = self.lora_linear_layer.down.weight.data.float()
+
+        if self.lora_linear_layer.network_alpha is not None:
+            w_up = w_up * self.lora_linear_layer.network_alpha / self.lora_linear_layer.rank
+
+        fused_weight = w_orig + torch.bmm(w_up[None, :], w_down[None, :])[0]
+        self.regular_linear_layer.weight.data = fused_weight.to(device=device, dtype=dtype)
+
+        # we can drop the lora layer now
+        self.lora_linear_layer = None
+
+        # offload the up and down matrices to CPU to not blow the memory
+        self.w_up = w_up.cpu()
+        self.w_down = w_down.cpu()
+
+    def _unfuse_lora(self):
+        if not (hasattr(self, "w_up") and hasattr(self, "w_down")):
+            return
+        logger.info(f"Unfusing LoRA weights for {self.__class__}")
+
+        fused_weight = self.regular_linear_layer.weight.data
+        dtype, device = fused_weight.dtype, fused_weight.device
+
+        self.w_up = self.w_up.to(device=device, dtype=dtype)
+        self.w_down = self.w_down.to(device, dtype=dtype)
+        unfused_weight = fused_weight - torch.bmm(self.w_up[None, :], self.w_down[None, :])[0]
+        self.regular_linear_layer.weight.data = unfused_weight.to(device=device, dtype=dtype)
```

Comment:
Actually the reason `unfuse` gives different results might be because we don't do the computation in full fp32 precision here. In the `_fuse_lora` function we do the computation in full fp32 precision, but here we don't I think. Can we try to make sure that:

```py
unfused_weight = fused_weight - torch.bmm(self.w_up[None, :], self.w_down[None, :])[0]
```

is always computed in full fp32 precision and only then we lower it potentially to fp16 dtype? 

Guess we could also easily check this by casting the whole model to fp32 before doing fuse and unfuse to check cc @apolinario

Reply:
This is a good hypothesis but unsure this is the case here

I'm testing out @sayakpaul's gists on many different LoRAs and this behavior seems to happen with some LoRAs and not others, some do give the identical image after the unload. Also, in the ones that don't - perceptually speaking - it seems like the LoRA effect is residually there

1. `davizca87/vulcan` → different image after unloading
2. `ostris/crayon_style_lora_sdxl` → same image after unloading
3. `davizca87/sun-flower` → different image after unloading (and also different than the image unlaoded in 1)
4. `TheLastBen/Papercut_SDXL`→ same image after unloading

Reply:
Failed cases ❌ 

`davizca87/sun-flower` seems to keep some of the sunflower vibe
|   original generation   |  with lora fused    |   after unfusing (residual effects from the LoRA?) |
|------|------|------|
|![image1](https://github.com/huggingface/diffusers/assets/788417/7794bee3-9e32-418d-9727-2226d05c38a7)|![image2](https://github.com/huggingface/diffusers/assets/788417/bc7d7420-f1be-420e-85c4-043aee13a578)|![image3](https://github.com/huggingface/diffusers/assets/788417/66207ac4-67e2-4958-955e-419cc0223e66)|

`davizca87/vulcan` seems to keep some of the vulcan style
|   original generation   |  with lora fused    |   after unfusing (residual effects from the LoRA?) |
|------|------|------|
|![image4](https://github.com/huggingface/diffusers/assets/788417/7794bee3-9e32-418d-9727-2226d05c38a7)|![image5](https://github.com/huggingface/diffusers/assets/788417/aa1e8313-0b1b-453e-9de6-8d7206a4e4f7)|![image6](https://github.com/huggingface/diffusers/assets/788417/7d150047-eb36-4350-8abd-a574e979cc18)|

Success cases ✅

`ostris/crayon_style_lora_sdxl` seems to produce a perceptually identical image 
|   original generation   |  with lora fused    |   after unfusing (same as original generation) |
|------|------|------|
|![image4](https://github.com/huggingface/diffusers/assets/788417/7794bee3-9e32-418d-9727-2226d05c38a7)|![image](https://github.com/huggingface/diffusers/assets/788417/242ff565-fd39-4674-857e-2a2f74f85c7a)|![image](https://github.com/huggingface/diffusers/assets/788417/fabe654b-63e2-4a2a-ae04-b11dc7a95923)|


`TheLastBen/Papercut_SDXL` and `nerijs/pixel-art-xl` also exhibit the same behavior

Reply:
I personally always worry about how numerical precision stems through a network and affect the end results. Have seen enough cases because of this to not sleep well at night. So, I will start with what @patrickvonplaten suggested. 

FWIW, though, [there's actually a fast test](https://github.com/huggingface/diffusers/blob/a76f2ad538e73b34d5fe7be08c8eb8ab38c7e90c/tests/models/test_lora_layers.py#L775) that ensures the following trip doesn't have side-effects:

`load_lora_weights() -> fuse_lora() -> unload_lora_weights()` gives you the outputs you would expect after doing `fuse_lora()`. 

Let me know if anything is unclear.

Reply:
Very nice! I guess an analogous test could be made to address what I reported above 

Namely asserting that the two "generate with no LoRA" are matching: 
`generate with no LoRA before unfusing` → `load_lora_weights()` →  `fuse_lora()` → `pipe.unfuse_lora()` → `generate with no LoRA unfusing`

This is the workflow I've reported above, the unfused unet still containing somewhat of a residue of the lora

Reply:
Should be fixed in: https://github.com/huggingface/diffusers/pull/4833 . Was a tricky issue that was caused by the patched text encoder LoRA layers being fully removed when doing `unload_lora` and therefore loosing their ability to unfuse. Hope it was ok to help out a bit here @sayakpaul

Reply:
Had started [`fuse-lora-pt2`](https://github.com/huggingface/diffusers/tree/fuse-lora-pt2) and mentioned in Slack that I am looking into it. But okay.